### PR TITLE
python[patch]: evaluators can return primitives

### PR DESCRIPTION
**Pull request by baskaryan**
```python
from langsmith import evaluate

def foo(run, example):
    return 0

def bar(run, example):
    return "long"

# removed, needs to be list of dict
# def baz(run, example):
#     return [0, 0.2, "how are ya"]

def app(inputs):
    return inputs

evaluate(app, data="Sample Dataset 3", evaluators=[foo, bar, baz])
```

### CODE REVIEW

**Comment by baskaryan** [comment]
how ^ example gets logged

Screenshot 2024-11-11 at 10 38 17 AM

**Comment by hinthornw** [comment]
The list return type seems unexpected to me - I wouldn't expect it to make separate keys for each value

**Comment by baskaryan** [comment]
> > > The list return type seems unexpected to me - I wouldn't expect it to make separate keys for each value
> > 
> > 
> > are we able to support list values as feedback?
> 
> Only dict or string it seems.
> 
> Maybe doing the _ix suffix is preferable. I'm not sure honestly. Probalby less surprising than just logging duplicates to the same key.
> 
> It just messes with the experiment averages (we'd be averaging over index 1 and over index 2 )

feel less strongly about list behavior either way, think main use case is supporting int/float/bool/str

**Comment by hinthornw** [comment]
Maybe let's land with the numeric and string value support but hold off on list behavior?

Maybe add support for list[evaluationresultlike] to make the "results" key not necessary

File: python/langsmith/evaluation/evaluator.py [hunk 1]
```diff
@@ -260,32 +259,46 @@ def _coerce_evaluation_results(
...
         self,
-        result: Union[EvaluationResult, EvaluationResults, dict],
+        result: Union[
+            EvaluationResult, EvaluationResults, dict, str, int, bool, float, list
+        ],
         source_run_id: uuid.UUID,
     ) -> Union[EvaluationResult, EvaluationResults]:
-        if isinstance(result, EvaluationResult):
+        if isinstance(result, (bool, float, int)):
+            result = {"score": result}
```
Comment (jakerachleff) [inline python/langsmith/evaluation/evaluator.py:278]: So if I have four categories that are numbers for some reason (I'm classifying college class levels and they're 100 level, 200 level, 300 level, etc), then I should do `str(value)` to explicitly use categorical scores?

File: python/tests/unit_tests/evaluation/test_runner.py [hunk 2]
```diff
@@ -347,11 +362,26 @@ async def score_value_first(run, example):
...
 
+    async def eval_float(run, example):
+        ordering_of_stuff.append("evaluate")
+        return 0.2
+
+    async def eval_str(run, example):
+        ordering_of_stuff.append("evaluate")
+        return "good"
+
+    async def eval_list(run, example):
```
Comment (jakerachleff) [inline python/tests/unit_tests/evaluation/test_runner.py:373]: maybe good to confirm that a list of ints, for example, doesn't work?

File: python/langsmith/evaluation/evaluator.py [hunk 3]
```diff
@@ -260,32 +259,46 @@ def _coerce_evaluation_results(
             cp = results.copy()
             cp["results"] = [
                 self._coerce_evaluation_result(r, source_run_id=source_run_id)
-                for r in results["results"]
+                for i, r in enumerate(results["results"])
```
Comment (hinthornw) [inline python/langsmith/evaluation/evaluator.py:262]: is `i` needed anymore?

**Review by hinthornw — APPROVED** [review]

**Pull request merged**